receiver/pprofreceiver/convert.go

```diff
@@ -0,0 +1,390 @@
+		// is no 1 to 1 mapping here.
+
+		// pprof.Profile.drop_frames
+		dropFramesIdx := getIdxForAttribute(transformCache, "drop_frames", src.DropFrames)
+		p.AttributeIndices().Append(dropFramesIdx)
+
+		// pprof.Profile.keep_frames
+		keepFramesIdx := getIdxForAttribute(transformCache, "keep_frames", src.KeepFrames)
+		p.AttributeIndices().Append(keepFramesIdx)
+
+		// pprof.Profile.time_nanos
+		p.SetTime(pcommon.Timestamp(src.TimeNanos))
+
+		// pprof.Profile.duration_nanos
+		p.SetDuration(pcommon.Timestamp(src.DurationNanos))
+
+		// pprof.Profile.period_type
+		p.PeriodType().SetTypeStrindex(getIdxForString(transformCache, src.PeriodType.Type))
+		p.PeriodType().SetUnitStrindex(getIdxForString(transformCache, src.PeriodType.Unit))
+
+		// pprof.Profile.period
+		p.SetPeriod(src.Period)
+
+		// pprof.Profile.comment
+		for _, c := range src.Comments {
+			idx := getIdxForString(transformCache, c)
+			p.CommentStrindices().Append(idx)
+		}
+
+		// pprof.Profile.default_sample_type
+		// As OTel pprofile uses a single Sample Type, it is implicit its default type.
+
+		// pprof.Profile.doc_url
+		docURLIdx := getIdxForAttribute(transformCache, "doc_url", src.DocURL)
+		p.AttributeIndices().Append(docURLIdx)
+	}
+
+	if err := dumpTransformCache(dst.Dictionary(), transformCache); err != nil {
+		return nil, err
+	}
+	return &dst, nil
+}
+
+// getIdxForString returns the corresponding index for the string.
+// If the string does not yet exist in the cache, it will be adedd.
+func getIdxForString(c cache, s string) int32 {
+	if idx, exists := c.stringTable[s]; exists {
+		return idx
+	}
+	c.lastStringTableIdx++
+	c.stringTable[s] = c.lastStringTableIdx
+	return c.lastStringTableIdx
+}
+
+// getIdxForAttribute returns the corresponding index for the attribute.
+// If the attribute does not yet exist in the cache, it will be added.
+func getIdxForAttribute(c cache, key string, value any) int32 {
+	return getIdxForAttributeWithUnit(c, key, "", value)
+}
+
+// getIdxForAttributeWithUnit returns the corresponding index for the attribute.
+// If the attribute does not yet exist in the cache, it will be added.
+func getIdxForAttributeWithUnit(c cache, key, unit string, value any) int32 {
+	keyStrIdx := getIdxForString(c, key)
+
+	unitStrIdx := noAttrUnit
+	if unit != "" {
+		unitStrIdx = getIdxForString(c, unit)
+	}
+
+	for attr, idx := range c.attributeTable {
+		if attr.keyStrIdx != keyStrIdx {
+			continue
+		}
+		if attr.unitStrIdx != unitStrIdx {
+			continue
+		}
+		if !isEqualValue(attr.value, value) {
+			continue
+		}
+		return idx
+	}
+
+	c.lastAttributeTableIdx++
+	c.attributeTable[attr{
+		keyStrIdx:  keyStrIdx,
+		value:      value,
+		unitStrIdx: unitStrIdx,
+	}] = c.lastAttributeTableIdx
+	return c.lastAttributeTableIdx
+}
+
+func isEqualValue(a, b any) bool {
+	return reflect.DeepEqual(a, b)
+}
+
+// initCache returns a supporting elements to construct pprofile.ProfilesDictionary.
+func initCache() cache {
```

Comment:
nit: IMO this is not quite a cache, it's more of lookup tables. Cache is typically something that is entirely transparent, with the main purpose of optimizing access. In particular, discarding the state of a cache should not have visible behavior except some operations taking longer. This is not the case here.

Reply:
Renamed `cache` to `lookupTables` with https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/42843/commits/4b1bed0b1e82272de16e4abe2769781bde58b09d.